File: android/src/main/kotlin/com/nanofuxion/tamertransports/LynxFetchModule.kt
package com.nanofuxion.tamertransports

import android.content.Context
import android.util.Base64
import android.util.Log
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException

class LynxFetchModule(context: Context) : LynxModule(context) {

    private val client = OkHttpClient()

    companion object {
        private const val TAG = "LynxFetchModule"
        private val BINARY_TYPES = setOf(
            "application/octet-stream",
            "application/pdf",
            "application/dns-message",
            "application/wasm",
            "image/",
            "audio/",
            "video/",
        )
    }

    private fun isBinaryContentType(contentType: String?): Boolean {
        if (contentType == null) return false
        return BINARY_TYPES.any { contentType.startsWith(it) }
    }

    @LynxMethod
    fun request(url: String, optionsJson: String, callback: Callback) {
        Log.d(TAG, "Fetch request: $url")
        try {
            val options = JSONObject(optionsJson)
            val method = options.optString("method", "GET").uppercase()
            val headers = options.optJSONObject("headers")
            val bodyStr = options.optString("body", "")
            val bodyBase64 = options.optString("bodyBase64", "")

            val builder = Request.Builder().url(url)

            if (headers != null) {
                val keys = headers.keys()
                while (keys.hasNext()) {
                    val key = keys.next()
                    builder.addHeader(key, headers.getString(key))
                }
            }

            val requestBody = when (method) {
                "GET", "HEAD" -> null
                else -> {
                    when {
                        bodyBase64.isNotEmpty() -> {
                            val contentType = headers?.optString("Content-Type") ?: "application/octet-stream"
                            val bytes = Base64.decode(bodyBase64, Base64.NO_WRAP)
                            okhttp3.RequestBody.create(contentType.toMediaType(), bytes)
                        }
                        else -> {
                            val contentType = headers?.optString("Content-Type") ?: "text/plain; charset=utf-8"
                            bodyStr.toRequestBody(contentType.toMediaType())
                        }
                    }
                }
            }
            builder.method(method, requestBody)

            val request = builder.build()

            client.newCall(request).execute().use { response ->
                val contentType = response.header("Content-Type")
                val rawBody = response.body?.bytes()
                val headersObj = JSONObject()
                response.headers.forEach { (name, value) ->
                    headersObj.put(name, value)
                }

                val result = JSONObject().apply {
                    put("ok", response.isSuccessful)
                    put("status", response.code)
                    put("statusText", response.message)
                    put("headers", headersObj)
                }
                if (rawBody != null && rawBody.isNotEmpty() && isBinaryContentType(contentType)) {
                    result.put("bodyBase64", Base64.encodeToString(rawBody, Base64.NO_WRAP))
                } else {
                    result.put("body", rawBody?.toString(Charsets.UTF_8) ?: "")
                }
                callback.invoke(result.toString())
            }
        } catch (e: IOException) {
            Log.e(TAG, "Fetch error: ${e.message}")
            val errorResult = JSONObject().apply { put("error", e.message ?: "Network error") }
            callback.invoke(errorResult.toString())
        } catch (e: Exception) {
            Log.e(TAG, "Fetch error: ${e.message}")
            val errorResult = JSONObject().apply { put("error", e.message ?: "Unknown error") }
            callback.invoke(errorResult.toString())
        }
    }
}

File: ios/tamertransports/tamertransports/Classes/LynxFetchModule.swift
import Foundation
import Lynx

@objcMembers
public final class LynxFetchModule: NSObject, LynxModule {

    @objc public static var name: String { "LynxFetchModule" }

    @objc public static var methodLookup: [String: String] {
        ["request": NSStringFromSelector(#selector(request(_:optionsJson:callback:)))]
    }

    private static let binaryPrefixes = [
        "application/octet-stream",
        "application/pdf",
        "application/dns-message",
        "application/wasm",
        "image/",
        "audio/",
        "video/",
    ]

    @objc public init(param: Any) { super.init() }
    @objc public override init() { super.init() }

    private func isBinaryContentType(_ contentType: String?) -> Bool {
        guard let ct = contentType else { return false }
        return Self.binaryPrefixes.contains { ct.hasPrefix($0) }
    }

    @objc func request(_ url: String, optionsJson: String, callback: @escaping (String) -> Void) {
        guard let urlObj = URL(string: url) else {
            callback(createJSONString(["error": "Invalid URL"]))
            return
        }

        var options: [String: Any] = [:]
        if let data = optionsJson.data(using: .utf8),
           let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            options = parsed
        }

        let method = (options["method"] as? String)?.uppercased() ?? "GET"
        let headers = options["headers"] as? [String: String] ?? [:]
        let bodyStr = options["body"] as? String
        let bodyBase64 = options["bodyBase64"] as? String

        var request = URLRequest(url: urlObj)
        request.httpMethod = method
        for (key, value) in headers { request.setValue(value, forHTTPHeaderField: key) }
        if !["GET", "HEAD"].contains(method) {
            if let base64 = bodyBase64, !base64.isEmpty, let bodyData = Data(base64Encoded: base64) {
                request.httpBody = bodyData
            } else if let body = bodyStr {
                request.httpBody = body.data(using: .utf8)
            }
        }

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                callback(self.createJSONString(["error": error.localizedDescription]))
                return
            }

            let httpResponse = response as? HTTPURLResponse
            let status = httpResponse?.statusCode ?? 0
            let statusText = HTTPURLResponse.localizedString(forStatusCode: status)
            let contentType = httpResponse?.value(forHTTPHeaderField: "Content-Type")

            var headersObj: [String: String] = [:]
            if let resp = httpResponse {
                for (key, value) in resp.allHeaderFields where key is String && value is String {
                    headersObj[key as! String] = value as? String
                }
            }

            var result: [String: Any] = [
                "ok": (200...299).contains(status),
                "status": status,
                "statusText": statusText,
                "headers": headersObj,
            ]
            if let data = data, !data.isEmpty, self.isBinaryContentType(contentType) {
                result["bodyBase64"] = data.base64EncodedString()
            } else {
                result["body"] = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
            }
            callback(self.createJSONString(result))
        }.resume()
    }

    private func createJSONString(_ dictionary: [String: Any]) -> String {
        (try? JSONSerialization.data(withJSONObject: dictionary)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
    }
}
